lib/utils/response.ts

```diff
@@ -19,8 +19,8 @@ const HTTP_STATUS_500 = 500 as HttpStatusCode;
  * Interface for a Response that can be returned by a middleware or
  * by the handlers.
  */
-export interface IResponse {
-  readonly kind: string;
+export interface IResponse<T> {
+  readonly kind: T;
```

Comment:
would it be useful to use an enum here (T extends MyEnum) with all the tags ? 
maybe to avoid typos when using the IResponses

Reply:
T can be anything (or extends string) ?

Reply:
looks like literal types can be a strings, numbers or booleans (see [discriminated unions](https://basarat.gitbooks.io/typescript/docs/types/discriminated-unions.html)) - if you misspell the type you should get a build error:

``` ts
interface IResponse<T> { 
    kind: T;
};

const a: IResponse<"A"> = {
    kind: "B", // error
};
```

Reply:
so shouldn't be 

```
export interface IResponse<T extends string>
```

moreover, I'm concerned about typos like

```
export function ContextMiddleware<T>(): IRequestMiddleware<"FooBar">
```

is there any check that 'FooBar' does exists indeed ?

Reply:
The point is not that `"FooBar"` exists, the point is that `IRequestMiddleware<"FooBar", T>` can only produce `IResponse<"FooBar">` responses.

Reply:
Try to play with the code and see by yourself that this refactoring doesn't lose any type safety.

Reply:
it looks like it works with const enum literals as discriminator. that would be good to keep intellisense autocomplete on kind(s) property.

Reply:
I'll merge this in the meantime, then you can do the PR with the enums